src/test/ui/reachable/expr_if.stderr

```diff
@@ -1,20 +1,3 @@
-error: unreachable block in `if` expression
-  --> $DIR/expr_if.rs:7:17
-   |
-LL |       if {return} {
-   |  _________------__^
-   | |         |
-   | |         any code following this expression is unreachable
-LL | |         println!("Hello, world!");
-LL | |     }
-   | |_____^ unreachable block in `if` expression
```

Comment:
@matthewjasper  The lint is triggered here

https://github.com/rust-lang/rust/blob/cf9bfdb8726c07974c8db93f70ee7213ee20c563/compiler/rustc_typeck/src/check/_match.rs#L274

And apparently it is necessary to check the previous expression (scrutinee or if condition)

https://github.com/rust-lang/rust/blob/cf9bfdb8726c07974c8db93f70ee7213ee20c563/compiler/rustc_typeck/src/check/_match.rs#L44

Taking this statement into consideration, the following lines

```rust
self.check_expr_has_type_or_error(cond_expr, self.tcx.types.bool, |_| {});
self.warn_if_unreachable(then_expr.hir_id, then_expr.span, "block in `if` expression");
```

Added at the begging of the `check_then_else` function, triggers the following ICE

```bash
error: internal compiler error: compiler/rustc_typeck/src/check/fn_ctxt/_impl.rs:306:25: while adjusting Expr { hir_id: HirId { owner: DefId(0:3785 ~ core[8580]::iter::adapters::filter::filter_fold), local_id: 21 }, kind: Path(Resolved(None, Path { span: library/core/src/iter/adapters/filter.rs:36:25: 36:34 (#0), res: Local(HirId { owner: DefId(0:3785 ~ core[8580]::iter::adapters::filter::filter_fold), local_id: 2 }), segments: [PathSegment { ident: predicate#0, hir_id: Some(HirId { owner: DefId(0:3785 ~ core[8580]::iter::adapters::filter::filter_fold), local_id: 20 }), res: Some(Local(HirId { owner: DefId(0:3785 ~ core[8580]::iter::adapters::filter::filter_fold), local_id: 2 })), args: None, infer_args: true }] })), attrs: ThinVec(None), span: library/core/src/iter/adapters/filter.rs:36:25: 36:34 (#0) }, can't compose [Borrow(Ref('_#0r, Mut { allow_two_phase_borrow: No })) -> &mut impl FnMut(&T) -> bool] and [Borrow(Ref('_#5r, Mut { allow_two_phase_borrow: No })) -> &mut impl FnMut(&T) -> bool]
```

Not sure why this is happening. Do you have any clues about this situation?